### PR TITLE
EOM-CCSD

### PR DESCRIPTION
**Pull request by amjames**
Adds modules for computing Hbar elements and doing a simple EOMCCSD computation. 

This can be very slow, but if you have access to Numpy 1.12 adding the kwarg optimize=True in the contractions for evaluating the sigma vectors will save a lot of time. 

The Davidson solver is not very robust and for some some cases where the number of roots requested involves the first one or two roots of a triplet set it will fail or converge to nonsensical complex roots. 

Any suggestions on how to improve the efficiency or the stability of the Davidson routine would be appreciated

### CODE REVIEW

**Comment by dgasmith** [comment]
@amjames Can you rebase in the latest changes and look into updating your headers to match? See the MP folder for examples.

**Comment by dgasmith** [comment]
I may have missed it, but is this ready to go? If so I can pull it in, ill probably rearrange the calls a bit- but the core looks good.

**Comment by amjames** [comment]
@dgasmith I still need to fix this up a little bit the results are not correct I have an error in my sigma equations. I will add references as well.

**Comment by dgasmith** [comment]
@amjames Poke poke.

**Comment by amjames** [comment]
@dgasmith acknowledging the poke. I've been swamped getting posters ready for WATOC. I'm flying Thursday and I'm going to be fixing this up on the plane.

**Comment by dgasmith** [comment]
Great, thanks AJ. I hope to get out another paper round this week.

**Comment by dgasmith** [comment]
@amjames We're moving a head with this project, any updates on this?

**Comment by dgasmith** [comment]
I think we are about to trip past my arbitrary deadline, will you be able to finish these up soon?

**Comment by dgasmith** [comment]
Are you planning to work on this? The deadline has technically passed for submissions.

**Comment by amjames** [comment]
@dgasmith I have the sigma equations checked and verified. Cleaning up the Davidson solver and adding refs then I will rebase/push. You can expect this ready to go today.

**Comment by dgasmith** [comment]
Great to know, thanks!

**Comment by amjames** [comment]
Annnnd, done.  I force pushed since my working implementation moved to the spin_free_CC folder. Let me know if you need me to change anything or add any other references.

**Comment by loriab** [comment]
Go ahead and add to [this file](https://github.com/psi4/psi4numpy/blob/master/tests/test_RI_CC.py) (2 lines; follow the pattern; use decorator if you need anything fancy) so that your script gets tested.

**Comment by amjames** [comment]
@loriab would that mean I want to do something like

`exe_py(workspace, tdir+'/spin_free_CC','EOM_CCSD')`

Since the script is in `Coupled-Cluster/spin_free_CC/`?

**Comment by loriab** [comment]
test shows up and passes on both. good to go.

File: Coupled-Cluster/spin_free_CC/EOM_CCSD.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+Spin orbital sigma equations can be found in I. Shavitt and R. J. Bartlett,
+"Many-Body Methods in Chemistry and Physics: MBPT and Coupled-Cluster Theory".
+Cambridge University Press, 2009.
+
+Special thanks to Ashutosh Kumar for Hbar components and help with spin
+adaptation.
+"""
+
+__authors__ = "Andrew M. James"
+__credits__ = ["T. Daniel Crawford", "Ashutosh Kumar"]
```
Comment (dgasmith) [inline Coupled-Cluster/spin_free_CC/EOM_CCSD.py]: Add yourself to the credits line as well.

File: Coupled-Cluster/spin_free_CC/EOM_CCSD.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+from helper_cc import helper_ccenergy
+from helper_cc import helper_cchbar
+from helper_cc import helper_cceom
+import psi4
+import time
+import numpy as np
+np.set_printoptions(precision=5, linewidth=200, threshold=200, suppress=True)
+
+if __name__ == '__main__':
```
Comment (dgasmith) [inline Coupled-Cluster/spin_free_CC/EOM_CCSD.py]: No need to wrap any of these scripts in a main as they are indeed just scripts.

File: Coupled-Cluster/spin_free_CC/EOM_CCSD.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    # EOM options
+    compare_psi4 = True
+    nroot = 3
+    nvec_per_root = 30
+    e_tol = 1.0e-6
+    max_iter = 80
```
Comment (dgasmith) [inline Coupled-Cluster/spin_free_CC/EOM_CCSD.py]: We usually do options after molecule spec.

File: Coupled-Cluster/spin_free_CC/EOM_CCSD.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            w = np.dot(S, alpha[:, j]) - theta[j] * np.dot(C, alpha[:, j])
+            # Precondition the residual vector to form a correction vector
+            q = w / (theta[j] - D[j])
+            # The correction vectors are added to the set of guesses after each
+            # iterations, so L the guess space dimension grows by nroot at each
+            # iteration
+            add_C.append(q)
+            de = abs(theta[j] - theta_old[j])
+            print(
+                "\tRoot {}: e = {:>20.12f} de = {:>20.12f} "
```
Comment (dgasmith) [inline Coupled-Cluster/spin_free_CC/EOM_CCSD.py]: Please do not use tabs, only spaces.

File: Coupled-Cluster/spin_free_CC/helper_cc.py [hunk 5]
```diff
@@ -1579,6 +1580,192 @@ def linresp(self):
 
 # End cclinresp class
 
+
+class helper_cceom(object):
```
Comment (dgasmith) [inline Coupled-Cluster/spin_free_CC/helper_cc.py]: Classes should be PascalCase not snake case.

File: Coupled-Cluster/spin_free_CC/helper_cc.py [hunk 6]
```diff
@@ -1579,6 +1580,192 @@ def linresp(self):
...
+        --------
+
+        >>> # Get some vectors as cols of a 2D numpy array and orthogonalize them
+        >>> c  = np.random.rand(eom.nsingles + eom.ndoubles, 2)
+        >>> c,  = np.linalg.qr(c)
+
+        >>> # Slice out the singles, doubles blocks of the first vector and reshape
+        >>> C1 = c[:,:nsingles].reshape(eom.ndocc, eom.nvir)
+        >>> C2 = c[:,nsingles:].reshape(eom.ndocc, eom.ndocc, eom.nvir, eom.nvir)
+        >>> S1 = eom.build_sigma1(C1, C2)
```
Comment (dgasmith) [inline Coupled-Cluster/spin_free_CC/helper_cc.py]: This class looks awesome, with docs! Can you add the original papers that you pull from in the `__init__` as well?

File: Coupled-Cluster/spin_free_CC/EOM_CCSD.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    # and we insert a guess vector with a 1 in the position corresponding to
+    # that single excitation and zeros everywhere else.
+    # This is a decent guess, more complicated one such as using CIS
+    # eigenvectors are more common in production level codes.
+    C_idx = D[:nov].argsort()[:nroot * 2]
+    C = np.eye(hbar_dim)[:, C_idx]
+
+    conv = False
+    eom_start = time.time()
+    for EOMCCSD_iter in range(0, max_iter + 1):
```
Comment (dgasmith) [inline Coupled-Cluster/spin_free_CC/EOM_CCSD.py]: Can you point out that this is the start of the Davidson iterations?

File: Coupled-Cluster/spin_free_CC/EOM_CCSD.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+            C1 = C[:nov, i].reshape(ndocc, nvir).copy()
+            C2 = C[nov:, i].reshape(ndocc, ndocc, nvir, nvir).copy()
+            S1 = cceom.build_sigma1(C1, C2)
+            S2 = cceom.build_sigma2(C1, C2)
+            S[:nov, i] += S1.flatten()
+            S[nov:, i] += S2.flatten()
+        # Build the subspace Hamiltonian
+        G = np.dot(C.T, S)
+        # Diagonalize it, and sort the eigenvector/eigenvalue pairs
+        THETA, alpha = np.linalg.eig(G)
```
Comment (dgasmith) [inline Coupled-Cluster/spin_free_CC/EOM_CCSD.py]: All caps variables should be reserved for GLOBAL vars.

File: Coupled-Cluster/spin_free_CC/EOM_CCSD.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    for EOMCCSD_iter in range(0, max_iter + 1):
+        # QR decomposition ensures that the columns of C are orthogonal
+        C, R = np.linalg.qr(C)
+        L = C.shape[1]
+        theta_old = theta[:nroot]
+        print("EOMCCSD: Iter # {:>6} L = {}".format(EOMCCSD_iter, L))
+        # Build up the matrix S, holding the products Hbar*C, aka sigma vectors
+        S = np.zeros_like(C)
+        for i in range(L):
+            C1 = C[:nov, i].reshape(ndocc, nvir).copy()
```
Comment (dgasmith) [inline Coupled-Cluster/spin_free_CC/EOM_CCSD.py]: Might want to use a different variable than C it took me a second you were not unpacking the orbitals.

File: Coupled-Cluster/spin_free_CC/helper_cc.py (hunk 5 above)
Comment (amjames) [reply Coupled-Cluster/spin_free_CC/helper_cc.py]: Got it, I was just matching the other helper classes above. Do you want me to change that on all the `helper_*` classes while I'm at it?

File: Coupled-Cluster/spin_free_CC/helper_cc.py (hunk 6 above)
Comment (amjames) [reply Coupled-Cluster/spin_free_CC/helper_cc.py]: Sure, you mean the docstring for `__init__` right?

Comment (amjames) [reply Coupled-Cluster/spin_free_CC/helper_cc.py]: Should that be a "References" section, "See Also" section, notes section, or something else?

File: Coupled-Cluster/spin_free_CC/helper_cc.py (hunk 5 above)
Comment (dgasmith) [reply Coupled-Cluster/spin_free_CC/helper_cc.py]: That would be great, yes.

File: Coupled-Cluster/spin_free_CC/helper_cc.py (hunk 6 above)
Comment (dgasmith) [reply Coupled-Cluster/spin_free_CC/helper_cc.py]: Yea, the `__init__` docstring. Please use the canonical `Notes` section for this docstring.

Comment (amjames) [reply Coupled-Cluster/spin_free_CC/helper_cc.py]: Do you want me to follow [this](https://github.com/numpy/numpy/blob/master/doc/example.py#L98-L108) style for the references (from the numpy docs examples) just put them under a `Notes` heading?

Comment (dgasmith) [reply Coupled-Cluster/spin_free_CC/helper_cc.py]: Just put them under notes. I think picking up references like that requires some special tech.

**Pull request merged**